### PR TITLE
[Feature] local participant persona avatar injection

### PR DESCRIPTION
**Pull request by iaulakh**
## Purpose

* Application should be able to inject avatar of local participant to UI Library
* groupcalloptions display name is always sent to SDK to display to remote participants
* Setupview, pip view, partiicpant view local user avatar displays name, avatar from injection if provided

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```
Settings to set Avatar:
Screen Shot 2022-04-13 at 8 20 31 PM

Setup screen persona view:
Screen Shot 2022-04-13 at 8 20 47 PM

Call screen persona view:

Screen Shot 2022-04-13 at 8 42 17 PM
Screen Shot 2022-04-13 at 8 22 30 PM
Screen Shot 2022-04-13 at 8 21 09 PM

### CODE REVIEW

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/PersonaManager.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.presentation.manager
+
+import com.azure.android.communication.ui.configuration.LocalParticipantConfiguration
+
+internal class PersonaManager(private val localParticipantConfiguration: LocalParticipantConfiguration?) {
```
Comment (JoshuaLai) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/PersonaManager.kt:8]: Just FYI on the iOS side I called this the `AvatarManager`

File: azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/CallLauncherActivity.kt [hunk 2]
```diff
@@ -21,16 +23,26 @@ import com.microsoft.appcenter.analytics.Analytics
...
     private lateinit var binding: ActivityCallLauncherBinding
     private val callLauncherViewModel: CallLauncherViewModel by viewModels()
     private val isTokenFunctionOptionSelected: String = "isTokenFunctionOptionSelected"
     private val isKotlinLauncherOptionSelected: String = "isKotlinLauncherOptionSelected"
+    private var localParticipantAvatarBitMap: Bitmap? = null
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+
+        Thread {
```
Comment (pavelprystinka) [inline azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/CallLauncherActivity.kt:39]: Could we move all this downloading image code to the dedicated method?

File: azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/CallLauncherActivity.kt [hunk 3]
```diff
@@ -21,16 +23,26 @@ import com.microsoft.appcenter.analytics.Analytics
...
+            val url =
+                URL("https://dt2sdf0db8zob.cloudfront.net/wp-content/uploads/2019/12/9-Best-Online-Avatars-and-How-to-Make-Your-Own-for-Free-image1-5.png")
```
Comment (pavelprystinka) [inline azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/CallLauncherActivity.kt:41]: Let's move it to buildConfigField that reads from local properties

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/CallComposite.java [hunk 4]
```diff
@@ -74,6 +75,43 @@ public void launch(final Context context, final GroupCallOptions groupCallOption
...
+     *
+     * @param context                       The android context used to start the Composite.
+     * @param groupCallOptions              The {@link GroupCallOptions} has parameters to
+     *                                      launch group call experience.
+     * @param localParticipantConfiguration The {@link LocalParticipantConfiguration} has parameters to
+     *                                      launch group call experience.
+     */
+    public void launch(final Context context,
+                       final GroupCallOptions groupCallOptions,
+                       final LocalParticipantConfiguration localParticipantConfiguration) {
```
Comment (pavelprystinka) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/CallComposite.java:103]: To me it is misleading and confusing to have a multiple option to provide a local participant name.

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/configuration/LocalParticipantConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+public class LocalParticipantConfiguration {
+    private final PersonaData personaData;
+
+    /**
+     * Create LocalParticipantConfiguration.
+     *
+     * @param personaData The {@link PersonaData};
+     * @see PersonaData
+     */
+    public LocalParticipantConfiguration(final PersonaData personaData) {
```
Comment (pavelprystinka) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/configuration/LocalParticipantConfiguration.java:40]: Is there a plan to have more data on the LocalParticipantConfiguration? It looks redundant to have this class with a single field

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/persona/PersonaData.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    private String name;
+    private ImageView.ScaleType scaleType = ImageView.ScaleType.FIT_XY;
+
+    /**
+     * Create PersonaData.
+     *
+     * @param image The {@link Bitmap};
+     * @see LocalParticipantConfiguration
+     */
+    public PersonaData(final Bitmap image) {
```
Comment (pavelprystinka) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/persona/PersonaData.java:42]: Parameter `image` is not clear. Should we call it `avatar` ?

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/configuration/LocalParticipantConfiguration.java (hunk 5 above)
Comment (JoshuaLai) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/configuration/LocalParticipantConfiguration.java:40]: There is a long term goal to allow developers to pass in things like: `title` org` kind of like what teams currently has

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/CallComposite.java (hunk 4 above)
Comment (JoshuaLai) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/CallComposite.java:103]: Yes we had this discussion on the iOS side as well. I think we need some way to give clarity to developers, this property is not going to be sent up to ACS, but values in the `GroupCallOptions` will be sent up to the server.

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/persona/PersonaData.java (hunk 6 above)
Comment (iaulakh) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/persona/PersonaData.java:42]: How about `avatarImage`?

Comment (iaulakh) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/persona/PersonaData.java:42]: changed to  `avatarBitmap`

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AvatarViewManager.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.presentation.manager
+
+import com.azure.android.communication.ui.configuration.CommunicationUILocalDataOptions
+
+internal class AvatarViewManager(private val communicationUILocalDataOptions: CommunicationUILocalDataOptions?) {
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AvatarViewManager.kt:8]: Any reason why we cannot make this public val?  Then the function below is not needed

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/utilities/BottomCellViewHolder.kt [hunk 8]
```diff
@@ -30,11 +30,19 @@ internal class BottomCellViewHolder(itemView: View) : RecyclerView.ViewHolder(it
 
     fun setCellData(bottomCellItem: BottomCellItem) {
         title.text = bottomCellItem.title
+        bottomCellItem.personaData?.let { personaData ->
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/utilities/BottomCellViewHolder.kt:33]: Can we do this instead:
`title.text = bottomCellItem.personaData?.renderedDisplayName ?: bottomCellItem.title`

**Review by phileo99 — CHANGES_REQUESTED** [review]

File: azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/SettingsActivity.kt [hunk 9]
```diff
@@ -131,16 +142,26 @@ class SettingsActivity : AppCompatActivity() {
             DEFAULT_LANGUAGE_VALUE
         )
     }
+
+    private fun updateRenderedDisplayNameSharedPreferences() {
+        sharedPreference.edit().putString(RENDERED_DISPLAY_NAME, renderDisplayNameTextView.text.toString()).apply()
+    }
+
+    private fun setAvatarSelection() {
```
Comment (ahammer) [inline azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/SettingsActivity.kt:150]: updateAvatarText

File: azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/SettingsActivity.kt [hunk 10]
```diff
@@ -40,6 +42,12 @@ class SettingsActivity : AppCompatActivity() {
         SettingsFeatures.initialize(this)
         supportedLanguages = LanguageCode.values().map { SettingsFeatures.displayLanguageName(it.toString()) }
         setLanguageInSharedPrefForFirstTime()
+        setAvatarSelection()
+    }
+
+    override fun onStop() {
+        super.onStop()
+        updateRenderedDisplayNameSharedPreferences()
```
Comment (ahammer) [inline azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/SettingsActivity.kt:50]: instead on onStop, do in onChangeListener

File: azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/SettingsActivity.kt [hunk 11]
```diff
@@ -131,16 +142,26 @@ class SettingsActivity : AppCompatActivity() {
             DEFAULT_LANGUAGE_VALUE
         )
     }
+
+    private fun updateRenderedDisplayNameSharedPreferences() {
```
Comment (ahammer) [inline azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/SettingsActivity.kt:146]: saveAvatarDisplayName()

File: azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/views/AvatarImageSelectionLinearlayout.kt [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        selectImageFromPreferences()
+
+        catImageButton.setOnClickListener {
+            selectImage(catImageButton)
+            setAvatarImagePreferences(catImageButton, R.drawable.image_cat.toString())
+        }
+
+        foxImageButton.setOnClickListener {
+            selectImage(foxImageButton)
+            setAvatarImagePreferences(foxImageButton, R.drawable.image_fox.toString())
```
[the remark this answers is not in the document]
Comment (iaulakh) [reply azure-communication-ui/azure-communication-ui-demo-app/src/main/java/com/azure/android/communication/ui/callingcompositedemoapp/views/AvatarImageSelectionLinearlayout.kt:46]: No Adam, We will include in documentation.

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/common/audiodevicelist/AudioDeviceListView.kt [hunk 13]
```diff
@@ -108,7 +108,8 @@ internal class AudioDeviceListView(
                     ),
                     null,
                     context.getString(R.string.azure_communication_ui_setup_view_audio_device_selected_accessibility_label),
-                    enabled = initialDevice == AudioDeviceSelectionStatus.RECEIVER_SELECTED
+                    enabled = initialDevice == AudioDeviceSelectionStatus.RECEIVER_SELECTED,
+                    null
```
Comment (ahammer) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/common/audiodevicelist/AudioDeviceListView.kt:112]: BottomCellItem should use Named params so we can leave these nulls out, and be a bit more clear about what we are passing in (too many params) can follow up in future PR.

Comment (iaulakh) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/common/audiodevicelist/AudioDeviceListView.kt:112]: will do that as part of refactoring, nice suggestion :)

**Review by phileo99 — APPROVED** [review]
tested, tested with emoji's also.
👍

**Review by lpeixin — APPROVED** [review]

**Pull request merged**